src/open_pulse/pulse_utils.cpp

```diff
@@ -0,0 +1,122 @@
+#include "pulse_utils.hpp"
+#include "zspmv.hpp"
+
+complex_t internal_expect_psi_csr(const py::array_t<complex_t>& data,
+                                  const py::array_t<int>& ind,
+                                  const py::array_t<int>& ptr,
+                                  const py::array_t<complex_t>& vec) {
+    auto data_raw = data.unchecked<1>();
+    auto vec_raw = vec.unchecked<1>();
+    auto ind_raw = ind.unchecked<1>();
+    auto ptr_raw = ptr.unchecked<1>();
+
+    auto nrows = vec.shape(0);
+    complex_t temp, expt = 0;
+
+    for (decltype(nrows) row = 0; row < nrows; row++) {
+        temp = 0;
+        auto vec_conj = std::conj(vec_raw[row]);
+        for (auto jj = ptr_raw[row]; jj < ptr_raw[row + 1]; jj++) {
+            temp += data_raw[jj] * vec_raw[ind_raw[jj]];
+        }
+        expt += vec_conj * temp;
+    }
+    return expt;
+}
+
+
+py::object expect_psi_csr(py::array_t<complex_t> data,
+                          py::array_t<int> ind,
+                          py::array_t<int> ptr,
+                          py::array_t<complex_t> vec,
+                          bool isherm){
+    complex_t expt = internal_expect_psi_csr(data, ind, ptr, vec);
+    if(isherm){
+        return py::cast(std::real(expt));
+    }
+    return py::cast(expt);
+}
+
+
+py::array_t<double> occ_probabilities(py::array_t<int> qubits,
+                                      py::array_t<complex_t> state,
+                                      py::list meas_ops){
+    auto meas_size = meas_ops.size();
+    py::array_t<double> probs(meas_size);
+    auto probs_raw = probs.mutable_unchecked<1>();
+    for(decltype(meas_size) i=0; i < meas_size; i++){
+        auto data = meas_ops[i].attr("data").attr("data").cast<py::array_t<complex_t>>();
+        auto ind = meas_ops[i].attr("data").attr("indices").cast<py::array_t<int>>();
+        auto ptr = meas_ops[i].attr("data").attr("indptr").cast<py::array_t<int>>();
+
+        probs_raw[i] = std::real(internal_expect_psi_csr(data, ind, ptr, state));
+    }
+
+    return probs;
+}
+
+void write_shots_memory(py::array_t<unsigned char> mem,
+                        py::array_t<unsigned int> mem_slots,
+                        py::array_t<double> probs,
+                        py::array_t<double> rand_vals)
+{
+    auto nrows = mem.shape(0);
+    auto nprobs = probs.shape(0);
+
+    unsigned char temp;
+
+    auto mem_raw = mem.mutable_unchecked<2>();
+    auto mem_slots_raw = mem_slots.unchecked<1>();
+    auto probs_raw = probs.unchecked<1>();
+    auto rand_vals_raw = rand_vals.unchecked<1>();
+
+    for(decltype(nrows) ii = 0; ii < nrows; ii++){
+        for(decltype(nprobs) jj = 0; jj < nprobs; jj++) {
+            temp = static_cast<unsigned char>(probs_raw[jj] > rand_vals_raw[nprobs*ii+jj]);
```

Comment:
... also, my other point with one single letter for the index variable is because I found clearer this expression:
```c++
rand_vals_raw[nprobs*i+j]);
```
than:
```c++
rand_vals_raw[nprobs*ii+jj]);
```

Reply:
I agree.